shell/packages/sandstorm-db/db.js

```diff
@@ -1875,6 +1891,77 @@ _.extend(SandstormDb.prototype, {
 
     return quota && user.storageUsage && user.storageUsage >= quota.storage * 1.2 && "outOfStorage";
   },
+
+  suspendIdentity: function (userId, suspension) {
+    this.collections.users.update({ _id: userId }, { $set: { suspended: suspension } });
+    this.collections.apiTokens.update({ "owner.user.identityId": userId },
+      { $set: { suspended: true } }, { multi: true });
+  },
+
+  unsuspendIdentity: function (userId) {
+    this.collections.users.update({ _id: userId }, { $unset: { suspended: 1 } });
+    this.collections.apiTokens.update({ "owner.user.identityId": userId },
+      { $unset: { suspended: true } }, { multi: true });
+  },
+
+  suspendAccount: function (userId, byAdminUserId, willDelete) {
+    const user = globalDb.collections.users.findOne({ _id: userId });
+    const suspension = {
+      timestamp: new Date(),
+      willDelete: willDelete || false,
+    };
+    if (byAdminUserId) {
+      suspension.admin = byAdminUserId;
+    } else {
+      suspension.voluntary = true;
+    }
+
+    this.collections.users.update({ _id: userId }, { $set: { suspended: suspension } });
+    this.collections.grains.update({ userId: userId }, { $set: { suspended: true } }, { multi: true });
+
+    delete suspension.willDelete;
+    // Only mark the parent account for deletion. This makes the query simpler later.
+
+    user.loginIdentities.forEach((identity) => {
+      this.suspendIdentity(identity.id, suspension);
+    });
+    // By default, suspend all login identities, but leave non-login identities alone.
+
+    if (byAdminUserId) {
+      // Force logout this user if suspended by an admin
```

Comment:
Maybe we should force logout for self-suspensions as well. Otherwise I'm worried the client-side code may be left in a state where it is still doing things in the background -- e.g. calling keepAlive() on grains that were open -- that we don't really expect to happen while suspended. I think being logged out is an unsurprising thing to happen when self-deleting.